config/credential_handler.rb

```diff
@@ -0,0 +1,150 @@
+require 'jwt'
+
+# This module is used in the Doorkeeper initializer for extracting client credentials upon
+# access token request.
+module CredentialHandler
+  # This method is called automatically by Doorkeeper upon client request for an access token.
+  def self.call(request)
+    grant_flow = request.parameters[:grant_type]
+    raise Doorkeeper::Errors::InvalidTokenStrategy unless grant_flow.present?
+
+    # Extract and handle credentials based on specified grant flow.
+    case grant_flow
+    when Doorkeeper::OAuth::AUTHORIZATION_CODE
+      # Simply extract the sent client id and secret - continued validation of these credentials
+      # as well as authorization code validation will happen after automatically for this flow. 
+      return request.parameters[:client_id], request.parameters[:client_secret]
+    when Doorkeeper::OAuth::CLIENT_CREDENTIALS
+      # This flow does not require an authorization code but instead requires JWT validation using JWK sets.
+      # This is not functionality supported by Doorkeeper intrinsically, so the custom validation is implemented here.
+      # Once the validation handled here passes, the client id and secret are sent along and no further validation is
+      # required before Doorkeeper sends an access token for this flow.
+      handle_client_credentials_flow(request)
+    else
+      # Unsupported grant type.
+      # NOTE: It should be impossible to get here with Doorkeeper's initial validation.
+      raise Doorkeeper::Errors::InvalidTokenStrategy
+    end
+  end
+
+  # Sara Alert supports the Smart on FHIR Backend Services protocol
+  # as described here: https://hl7.org/fhir/uv/bulkdata/authorization/index.html
+  # This requires the use client_credentials grant flow with JWT assertions.
+  # This method handles authorizing the client by decoding the JWT assertion for the supported client credentials flow. 
+  def self.handle_client_credentials_flow(request)
+    # Extract the shared client ID.
+    client_id = request.parameters[:client_id]
+    raise_missing_param_error('client_id') unless client_id.present?
```

Comment:
Why is client_id required? client_id can be obtained from decode (without verifying) JWT token.
decoded_assertion = JWT.decode signed_token, nil, false
client_id = decoded_assertion.iss